File: src/main/java/burrow/core/furniture/DependencyTree.java
package burrow.core.furniture;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Consumer;

public final class DependencyTree {
    private final Node root = new Node(null);

    @NotNull
    public Node getRoot() {
        return root;
    }

    public void resolve(@NotNull final Consumer<Furniture> consumer) {
        new ResolveRoutine().resolveNode(root, consumer);
    }

    @NotNull
    public Node createNode(@NotNull final Furniture furniture) {
        return new Node(furniture);
    }

    public static final class Node {
        private final List<Node> children = new ArrayList<>();
        private final Furniture furniture;

        public Node(Furniture furniture) {this.furniture = furniture;}

        public void add(@NotNull final Node child) {
            children.add(child);
        }
    }

    private static final class ResolveRoutine {
        private final Set<Node> resolvedNode = new HashSet<>();

        private ResolveRoutine() {
            resolvedNode.add(null);
        }

        private void resolveNode(
            @NotNull final Node node,
            @NotNull final Consumer<Furniture> consumer
        ) {
            if (resolvedNode.contains(node)) {
                return;
            }

            for (final var child : node.children) {
                resolveNode(child, consumer);
            }

            if (node.furniture != null) {
                consumer.accept(node.furniture);
            }

            resolvedNode.add(node);
        }
    }
}

File: src/main/kotlin/kernel/furnishing/DependencyTree.kt
package burrow.kernel.furnishing

open class DependencyTree<T> {
    open class Node<T>(val element: T?) {
        val children = mutableListOf<Node<T>>()

        fun add(element: T) {
            children.add(Node(element))
        }
    }

    private class ResolveRoutine<T> {
        private val resolvedNodes = mutableSetOf<Node<T>>()

        fun resolve(node: Node<T>, consumer: (T) -> Unit) {
            if (resolvedNodes.add(node)) {
                node.children.forEach { resolve(it, consumer) }
                node.element?.let(consumer)
            }
        }
    }

    val root = Node<T>(null)

    fun resolve(resolver: (T) -> Unit) {
        ResolveRoutine<T>().resolve(root, resolver)
    }
}

class FurnishingDependencyTree : DependencyTree<Furnishing>()